_posts/2022-09-07-storage-write-removal-before-conditional-termination.md

```diff
@@ -0,0 +1,133 @@
+
+When using the legacy code generator, the Yul optimizer step responsible for the bug will be run on each inline assembly block in isolation and hence the above conditions have to occur inside one inline assembly block. While still possible, this makes the occurrence of the bug with legacy code generation less likely (see the section of Technical Details below for examples of affected assembly snippets).
+
+However, the new via-IR code generation translates the entire contract to Yul and runs the optimizer on all of it. In this case, the bug can be triggered more easily. The following is a minimal example:
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1; // This write used to be removed by the yul optimizer.
+		g();
+		x = 2;
+	}
+	function g() internal {
+		// The important aspect of this function is that it can both return to the caller and terminate the transaction.
+		// The bug will show its effects in the cases in which it terminates the transaction, since the state change above will be missing.
+		if (msg.data.length > 4) return;
+		assembly { return(0,0) }
+	}
+}
+```
+
+Compiling the above via-IR with enabled optimizer will result ``f()`` without additional calldata to incorrectly return without modifying ``x``.
+
+For a concrete contract, the bug may still be prevented, in case the optimizer inlines the function before running the problematic store removal step, e.g. the following is almost
+identical, but unaffected, since ``g()`` will be inlined (in the snippet above the ``return;`` prevents easy inlining):
+
+```
+contract C {
+	uint public x;
+	function f() public {
+		x = 1;
+		g();
+		x = 2;
+	}
+	function g() internal {
+		if (msg.data.length <= 4)
+			assembly { return(0,0) }
+	}
+}
+```
+
+
+## Technical Details
+
+On the Yul level the issue manifests if the following sequence occurs:
+
+- A storage write occurs (e.g. ``sstore(0, 1)``).
+- A user-defined Yul function is called and this function both has a control flow path terminates the transaction and a control flow path out of the function.
+- Another storage write occurs that makes the first storage write above redundant (e.g. ``sstore(0, 2)``).
+
+In this situation, the optimizer will treat the first storage write as globally redundant, despite the fact that the called Yul function may successfully terminate the transaction, in which case the first storage write is not redundant at all.
+
+The following is a full example of a vulnerable Yul snippet:
+
+```
+{
+  function f() {
+    if gt(calldatasize(), 4) { leave }
+    return(0,0)
+  }
+  sstore(0, 1) // This sstore will incorrectly be removed.
+  f()
+  sstore(0, 2)
+}
+```
+
+At the call of ``f()``, the first ``sstore`` should already be marked as ``used`` by the optimizer, since the control flow path of ``f()`` reaching the ``return(0,0)`` relies on it.
```

Comment:
Maybe we should add some explanation about how the steps works so that it's clear what `used` means in this context? I.e. the way I described it in https://github.com/ethereum/solidity/pull/13479#discussion_r963099289. Because technically the value might not be used - it's just exposed to the outside and can *potentially* be read.

Reply:
Only reading the comment in the other PR now - but the new version of the blog post hopefully explains the optimizer step better already.

Reply:
Yeah, I think it does explain it well now.